src/fm.py

```diff
@@ -33,5 +33,172 @@ def main():
+        lo = lower(a, genome, sa, lo, hi, offset)
+        hi = upper(a, genome, sa, lo, hi, offset)
+    for sol in sa[lo:hi]:
+        yield sol+1
+    return
+
+def getSuffixes(x):
+    """
+    Gets all suffixes from a string
+    """
+    suffixes = [x[i:] for i in range(0, len(x))] 
+    # print("list of unordered suffixes: ",suffixes) 
+    return suffixes
+
+def radix_sort(lst: list[memoryview]):
+    # print("Radix sort", len(lst))
+    maximum = max(lst, key = len)
+
+    for place in range(len(maximum),0, -1):
+        lst = counting_sort(lst, place - 1)
+    
+    lst = [len(maximum) - len(suffix) for suffix in lst]
+    return lst
+            
+
+def counting_sort(lst: list[memoryview], place):
+    maximum = max(lst, key = len)
+    counts = dict.fromkeys(["$",*sorted(maximum)],[])
+    for string_index in range(len(lst)):
+        if place >= len(lst[string_index]):
+            counts["$"] = [*counts["$"], lst[string_index]]
+            
+        else:
+            counts[lst[string_index][place]] = [*counts[lst[string_index][place]], lst[string_index]]
+
+    ordered = []
+    for key in counts:
+        ordered += counts[key]
+    return ordered
+
+def build_rank_table(x, alphadic, bwt):
+    counts = {a: 0 for a in alphadic}
+    table = [[0 for _ in alphadic] for _ in range(0, len(bwt)+1)]
+
+    for i in range(1, len(bwt)+1):        
+        for j in range(0, len(alphadic)):
+            table[i][j] = table[i-1][j]
+        
+        bwtValue = bwt[i-1]
+        c = chr(x[bwtValue])
+
+        # Add the count
+        counts[c] += 1
+
+        index = alphadic.get(c)
+        table[i][index] += 1
+
+    return table, counts
+
+def getFirstIndexList(x, f, alphadic):
+    firstIndexList = {a: -1 for a in alphadic}
+    indexesFound = 0
+
+    for i, xIndex in enumerate(f):   
+        c = chr(x[xIndex])
+        if firstIndexList.get(c) < 0:
+            firstIndexList[c] = i
+            indexesFound += 1
+
+            if indexesFound >= len(alphadic):
+                return firstIndexList
+    
+    return firstIndexList
+
+
+def getrank(alphadic, index, c, rank_table):
+    return rank_table[index][alphadic.get(c)]
+
+def searchPattern(x, p, f, l, rank_table, firstIndexList, countTable, alphadic):
+    if p == "":
+        return
+    
+    if len(p) == 1:
+        firstMatch = firstIndexList.get(p[0])
+        for i in range(0, countTable.get(p[0])):
+            yield f[i + firstMatch]
+        return
+
+    firstRank = -1
+    nRanks = 0
+
+    # First case
+    if True:
+        c = p[len(p)-1] # b
+        upperBound = firstIndexList.get(c)
+        lowerBound = upperBound+countTable.get(c)
+
+        for j in range(upperBound, lowerBound):
+            print (chr(x[l[j]]), p[len(p)-2], sep=" == ")
+            if chr(x[l[j]]) == p[len(p)-2]:
+                if firstRank < 0:
+                    firstRank = getrank(alphadic, j, p[len(p)-2], rank_table)
+                nRanks += 1
```

Comment:
I don't know what you are doing here, but I know it isn't necessary. The FM-index is one of the simplest algorithms on the planet to implement (but not, by far, to understand), and there really isn't any special case.

Building the suffix array and the tables might not be trivial, but the search is. The general case that updates indices

```python
  i = C[a] + O[a, i]
```

handles *everything*.

This is literally my complete implementation, searching and reporting in 7 lines (not including blank and comment):

```python
        left, right = 0, len(sa)
        for a in reversed(p):
            left = ctab[a] + otab[a, left]
            right = ctab[a] + otab[a, right]
            if left >= right: return  # no matches

        # Report the matches
        for i in range(left, right):
            yield sa[i]
```